services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/enhanced/dynamodb/TableSchema.java

```diff
@@ -131,12 +132,50 @@
      * If attributes are missing from the map, that will not cause an error, however if attributes are found in the
      * map which the mapper does not know how to map, an exception will be thrown.
      *
+     * <p>
+     * If all attribute values in the attributeMap are null, null will be returned. Use {@link #mapToItem(Map, boolean)}
+     * instead if you need to preserve empty object.
+     *
+     * <p>
+     * API Implementors Note:
+     * <p>
+     * {@link #mapToItem(Map, boolean)} should be implemented if {@code preserveEmptyObject} behavior is desired.
+     *
      * @param attributeMap A map of String to {@link AttributeValue} that contains all the raw attributes to map.
      * @return A new instance of a Java object with all the attributes mapped onto it.
      * @throws IllegalArgumentException if any attributes in the map could not be mapped onto the new model object.
+     * @see #mapToItem(Map, boolean)
      */
     T mapToItem(Map<String, AttributeValue> attributeMap);
 
+    /**
+     * Takes a raw DynamoDb SDK representation of a record in a table and maps it to a Java object. A new object is
+     * created to fulfil this operation.
+     * <p>
+     * If attributes are missing from the map, that will not cause an error, however if attributes are found in the
+     * map which the mapper does not know how to map, an exception will be thrown.
+     *
+     * <p>
+     * In the scenario where all attribute values in the map are null, it will return null if {@code preserveEmptyObject}
+     * is true. If it's false, an empty object will be returned.
+     *
+     * <p>
+     * Note that {@code preserveEmptyObject} only applies to the top level Java object, if it has nested "empty" objects, they
+     * will be mapped as null. You can use {@link DynamoDbPreserveEmptyObject} to configure this behavior for nested objects.
+     *
+     * @param attributeMap A map of String to {@link AttributeValue} that contains all the raw attributes to map.
+     * @param preserveEmptyObject whether to initialize this Java object as empty class if all fields are null
+     * @return A new instance of a Java object with all the attributes mapped onto it.
+     * @throws IllegalArgumentException if any attributes in the map could not be mapped onto the new model object.
+     * @see #mapToItem(Map)
+     */
+    default T mapToItem(Map<String, AttributeValue> attributeMap, boolean preserveEmptyObject) {
+        if (preserveEmptyObject) {
+            throw new UnsupportedOperationException();
```

Comment:
I think there should be a descriptive message here, as one situation this could be encountered is someone attempting to use an existing [external] implementation of TableSchema and I think a message letting them know why their implementation cannot be used the way they are trying to use it (presumably by setting preserveEmptyObject to true) would be helpful.

Reply:
Yeah, added